bridge-balance-difference/src/agent.js

```diff
@@ -0,0 +1,122 @@
+const {
+  Finding,
+  FindingSeverity,
+  FindingType,
+  ethers,
+} = require('forta-agent');
+const ARIMA = require('arima');
+const { contract1, contract2, tokens: t } = require('../bot-config.json');
+
+const INTERVAL = 3600; // 1 hour
+
+const secondsPerYear = 60 * 60 * 24 * 365;
+const periodsPerYear = Math.floor(secondsPerYear / INTERVAL);
+
+const arima = new ARIMA({
+  p: 1,
+  d: 0,
+  q: 1,
+  verbose: false,
+});
+
+const ABI = [
+  'function balanceOf(address) public view returns (uint256)',
+  'function decimals() external view returns (uint8)',
+];
+
+const provider1 = new ethers.providers.JsonRpcProvider(contract1.rpcUrl);
+const provider2 = new ethers.providers.JsonRpcProvider(contract2.rpcUrl);
+
+// Each token has 2 contracts
+const tokens = t.map((token) => ({
+  timeSeries: [],
+  decimals: 0,
+  tokenContract1: new ethers.Contract(token.address1, ABI, provider1),
+  tokenContract2: new ethers.Contract(token.address2, ABI, provider2),
+}));
+
+let chainId1;
+let chainId2;
+let lastTimestamp = 0;
+
+function provideInitialize(prov1, prov2, tokensObj) {
+  return async function initialize() {
+    chainId1 = (await prov1.getNetwork()).chainId;
+    chainId2 = (await prov2.getNetwork()).chainId;
+
+    const decimals = await Promise.all(
+      tokensObj.map(async (token) => token.tokenContract1.decimals()),
+    );
+
+    decimals.forEach((dec, i) => {
+      tokens[i].decimals = dec;
+    });
+  };
+}
+
+function provideHandleBlock(tokensData) {
+  return async function handleBlock(blockEvent) {
+    let findings = [];
+    const { timestamp } = blockEvent.block;
+
+    // Check the price once every INTERVAL
+    if (timestamp < lastTimestamp + INTERVAL) return findings;
+
+    findings = await Promise.all(tokensData.map(async (token) => {
+      const {
+        tokenContract1,
+        tokenContract2,
+        decimals,
+        timeSeries,
+      } = token;
+      let tempFinding = null;
+
+      // Get the token balance for both bridge contracts
+      const balance1 = await tokenContract1.balanceOf(contract1.address);
+      const balance2 = await tokenContract2.balanceOf(contract2.address);
+
+      // Keep only 2 decimals because ARIMA doesn't work well with a lot of decimals
+      let difference = ethers.utils.formatUnits(balance1.sub(balance2), decimals);
+      difference = parseFloat((+difference).toFixed(2));
+
+      if (timeSeries.length > 10) {
+        arima.train(timeSeries);
+        const [pred, err] = arima.predict(1).flat();
+
+        // Calculate the 95% confidence interval
+        const high = pred + 1.96 * Math.sqrt(err);
+
+        if (difference > high) {
```

Comment:
does there need to be a time component as well where the assets may take some time to transfer?

Reply:
I think we can either make the interval much larger than the time to transfer or we can introduce a lag, so we will compare the balance on chain 1 at time X with the balance on chain 2 at time X - lag. What do you think is better?